### PR TITLE
Sync the callback API with new tf.keras API

### PR DESCRIPTION
**Pull request by mkaze**
### Summary
This PR syncs callback API with [the new `tf.keras` API](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/callbacks.py#L175).

Modifications:
- A new helper function, `_call_batch_hook`, has been introduced to handle batch-related callbacks methods in different modes (train/test/prediction).
- `on_batch_end` and `on_batch_begin` has been refactored to call `_call_batch_hook` instead.
- Methods `on_train_batch_begin` and `on_train_batch_end` have been introduced specifically for callbacks batch-related methods in training mode.

I have also done some small formatting and style fixes.

Further, I must mention that I think there is a small error in TF implementation: [in this line](https://github.com/tensorflow/tensorflow/blob/320dc4913f94a679790d90f5c577037276aa9599/tensorflow/python/keras/callbacks.py#L187) the `_t_enter_batch` attribute, which measure the batch processing time, is set at the beginning of the method. However, I think this is wrong (since in that case the execution time of `on_batch_begin` would also be included) and as it is [currently implemented in Keras](https://github.com/keras-team/keras/blob/7bee9a18d83899ce6dfd50c4883afc678139f4ad/keras/callbacks.py#L99), it must be done at the end of the method. So in this PR I would set this attribute at the end.

### Related Issues
#11771 

### PR Overview

- [?] This PR requires new unit tests [y/n] (make sure tests are included)
- [?] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [y] This PR is backwards compatible [y/n]
- [y] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
Thank you very much for the pull request. @fchollet will have to be one of the reviewers since there are API changes.

File: keras/callbacks.py [hunk 1]
```diff
@@ -214,10 +253,7 @@ class BaseLogger(Callback):
     """
 
     def __init__(self, stateful_metrics=None):
-        if stateful_metrics:
-            self.stateful_metrics = set(stateful_metrics)
-        else:
-            self.stateful_metrics = set()
```
Comment (gabrieldemarmiesse) [inline keras/callbacks.py]: Could you revert the changes unrelated to the sync of the callback api with tf.keras? You can create another pull request for them if you want. It's to make it easier to review and discuss each of your changes. Thanks.

File: keras/callbacks.py [hunk 2]
```diff
@@ -52,74 +58,99 @@ def set_model(self, model):
...
+        for callback in self.callbacks:
+            batch_hook = getattr(callback, hook_name)
+            batch_hook(batch, logs)
+        self._delta_ts[hook_name].append(time.time() - t_before_callbacks)
+
+        delta_t_median = np.median(self._delta_ts[hook_name])
+        if (self._delta_t_batch > 0. and
+           delta_t_median > 0.95 * self._delta_t_batch and
+           delta_t_median > 0.1):
+            logging.warning(
```
Comment (gabrieldemarmiesse) [inline keras/callbacks.py]: `logging` is not imported. Did you mean `warnings.warn()` ?

File: keras/callbacks.py [hunk 3]
```diff
@@ -52,74 +58,99 @@ def set_model(self, model):
...
+                'Method (%s) is slow compared '
+                'to the batch update (%f). Check your callbacks.', hook_name,
+                delta_t_median)
+        if hook == 'begin':
+            self._t_enter_batch = time.time()
+
+    def _call_begin_hook(self, mode):
+        """Helper function for on_{train|test|predict}_begin methods."""
+        # only train mode is supported for now
```
Comment (gabrieldemarmiesse) [inline keras/callbacks.py]: Please raise a `NotIplementedError` when the use case is not supported yet.

File: keras/callbacks.py [hunk 4]
```diff
@@ -52,74 +58,99 @@ def set_model(self, model):
         for callback in self.callbacks:
             callback.set_model(model)
 
-    def on_epoch_begin(self, epoch, logs=None):
+    def _call_batch_hook(self, mode, hook, batch, logs=None):
+        """Helper function for all batch_{begin | end} methods."""
+        # only train mode is supported for now
+        if mode != 'train':
+            return
```
Comment (gabrieldemarmiesse) [inline keras/callbacks.py]: `raise NotImplementedError` may be more appropriate.

**Comment by mkaze** [comment]
@gabrieldemarmiesse Thanks for the reviews and sorry for the delay. I applied the followings according to your reviews:

- Fixed `warnings.warn`
- Raised `NotImplementedError` for unsupported modes
- Revert back unrelated changes

Further, I have added a check in `_call_batch_hook` to make sure `_t_enter_batch` is initialized properly when `hook == end`.

**Comment by mkaze** [comment]
@omalleyt12 Thanks for the review. Added the hooks for test and predict. Yet I need to modify the training methods as well as add the tests.

**Comment by mkaze** [comment]
@omalleyt12 @gabrieldemarmiesse @fchollet OK, finally it is finished and is ready for your reviews! The callback API is synced and also the callback support has been added to `fit`, `predict`, `evaluate`, `fit_generator`, `predict_generator` and `evaluate_generator` methods. Further, the tests have also been added for both array-based as well as the generator-based methods. Additionally, I would like to update the [callbacks documentation for creating new callbacks](https://keras.io/callbacks/#create-a-callback) since I think it is too brief and now with this PR, it surely needs an update. However, I would do it in another PR after this one has been merged. 

@omalleyt12 Thanks a lot. Without using your implementation as a guide it was so hard for me to do this. Just one question: I think there is a mistake in the docstring of `on_predict_batch_{begin|end}` methods. Actually, the `logs` argument contains the predictions and not the metrics on the given batch. Therefore it must be:

    logs: dict, predictions for this batch.

instead of:

    logs: dict, metric results for this batch.

Am I right? If that's the case, then I need to fix this as well.

**Comment by mkaze** [comment]
@fchollet Do we need to add the support for callbacks to the `predict_classes` and `predict_proba` of sequential methods as well? They essentially call `predict` under the hood so it would be easy to this.

**Comment by mkaze** [comment]
@gabrieldemarmiesse @omalleyt12 @fchollet No news?! It seems you are all enjoying your holidays, right? Or is it something else?

**Comment by gabrieldemarmiesse** [comment]
@mkaze , it is indeed the holiday period, also, remember that this is an open source project, we do it in our free time, so we can't review every day. (and @fchollet is quite busy with TF2.0). A pull request which is big is also longer to review , actually, often exponentially to the number of lines. So we need to find the free time to sit ~2 hours in front of our computers to review. And there are many pull requests like that. (I'm not even mentioning the number of issues to read).

Please be patient. Sorry for the wait and thank you for your work, it's really useful (it's actually a very important sync which must be done).

**Comment by mkaze** [comment]
@gabrieldemarmiesse Thanks for the reply. You are completely right. This a big PR and it takes time to review it. I just saw some of the other PRs have been merged during past week and therefore wondered maybe there is a problem with my PR or maybe the plan for the sync has changed or something like that. That's why I just asked to know.

By the way, happy holidays and new year! And thanks for your efforts!

**Comment by fchollet** [comment]
@omalleyt12 Tom, does this still look good to you? Should we merge?

**Comment by omalleyt12** [comment]
> @omalleyt12 Tom, does this still look good to you? Should we merge?

Looks good to me!
